### PR TITLE
Made changes to sample page so it renders when no internet connection

### LINKED ISSUE #1212: Sample page doesn't render without internet connection
## Description

When a user selects a sample, the webview will fail to render if they have no internet connection. This is because we attempt to fetch the samples README.

`const response: any = await Axios.get(options.sample.readme);`

## Expected Behavior

We should either throw an error after a certain timeout value, or we should render the page without the README and let the user know we were unable to retrieve it.

## Actual Behavior

Sample webview fails to render.
## Possible Fix

On the `Axios.get()` request, we could pass a timeout value as one of the options.
## Steps to Reproduce

1. Open up Home page
2. Disconnect from internet
3. Click on FabCar or Commercial Paper sample

## Existing issues

- [ ] [Stack Overflow issues](https://stackoverflow.com/questions/tagged/ibp-vscode-extension)
- [ ] [GitHub Issues](https://github.com/IBM-Blockchain/blockchain-vscode-extension/issues)

## Your Environment

* IBM Blockchain Platform Extension Version:
* Visual Studio Code version:
* Environment name and version (e.g. Node.js v8.12.0, npm v6.4.1):
* Operating System and version:
* Link to your project:

### PR DESCRIPTION
**Pull request by toggle007**
Closes #1212 

Signed-off-by: Akshat Shah <akshatshah@akshats-mbp.hursley.uk.ibm.com>

### CODE REVIEW

File: src/webview/SampleView.ts [hunk 1]
```diff
@@ -135,14 +136,34 @@ export class SampleView extends View {
...
+        let validInternetConnection: boolean = true;
+        const outputAdapter: VSCodeBlockchainOutputAdapter = VSCodeBlockchainOutputAdapter.instance();
         const templatePath: string = path.join(__dirname, '..', '..', '..', 'templates', 'SampleView.ejs');
 
         const converter: showdown.Converter = new showdown.Converter();
-        const response: any = await Axios.get(options.sample.readme);
-        const text: string = response.data;
-        const html: string = converter.makeHtml(text);
-        options.html = html;
+        const repositoryRegistry: RepositoryRegistry = RepositoryRegistry.instance();
```
Comment (cazfletch) [inline src/webview/SampleView.ts]: you don't need to do this unless you don't have an internet connection

File: src/webview/SampleView.ts [hunk 2]
```diff
@@ -135,14 +136,34 @@ export class SampleView extends View {
...
+            options.html = html;
+        } catch (error) {
+            validInternetConnection = false;
+        }
+
+        if (!validInternetConnection) {
+            if (repoIsCloned) {
+                const clonedRepo: RepositoryRegistryEntry = repositoryRegistry.get(options.repositoryName);
+                const repoPath: string = clonedRepo.path;
+                const readMe: string = await fs.readFileSync(path.join(repoPath, 'README.md'), 'utf8');
```
Comment (cazfletch) [inline src/webview/SampleView.ts]: the readme might not always be at the top level, you need to get the readme property to work out where it is

File: test/webview/SampleView.test.ts [hunk 3]
```diff
@@ -233,6 +233,22 @@ describe('SampleView', () => {
             samplePageHtml.should.contain(`<button disabled class="open-button" onclick="openFile('applications','TypeScript Application')">Open Locally</button>`); // Disabled open button
         });
 
+        it('should display everything excluding the readme when there is no internet connection and the repository has not been cloned', async () => {
+            const outputAdapterSpy: sinon.SinonSpy = mySandBox.spy(VSCodeBlockchainOutputAdapter.instance(), 'log');
+            const repositoryRegistryExistsStub: sinon.SinonStub = mySandBox.stub(RepositoryRegistry.prototype, 'exists').returns(false);
```
Comment (cazfletch) [inline test/webview/SampleView.test.ts]: you don't need to stub this just make sure there is nothing in the registry

File: test/webview/SampleView.test.ts [hunk 4]
```diff
@@ -273,6 +289,35 @@ describe('SampleView', () => {
             samplePageHtml.should.contain(`<button class="open-button" onclick="openFile('applications','JavaScript Application')">Open Locally</button>`); // Disabled open button
             samplePageHtml.should.contain(`<button class="open-button" onclick="openFile('applications','TypeScript Application')">Open Locally</button>`); // Disabled open button
         });
+
+        it('should get correct html when cloned and there is no internet connection', async () => {
+            const repositoryRegistryExistsStub: sinon.SinonStub = mySandBox.stub(RepositoryRegistry.prototype, 'exists').returns(true);
```
Comment (cazfletch) [inline test/webview/SampleView.test.ts]: again don't need to stub this just add something to the registry

File: src/webview/SampleView.ts (hunk 1 above)
Comment (toggle007) [reply src/webview/SampleView.ts]: Fixed - moved it into the if statement once it checks that there is no internet connection.

File: src/webview/SampleView.ts (hunk 2 above)
Comment (toggle007) [reply src/webview/SampleView.ts]: Fixed

File: test/webview/SampleView.test.ts (hunk 3 above)
Comment (toggle007) [reply test/webview/SampleView.test.ts]: Fixed

File: test/webview/SampleView.test.ts (hunk 4 above)
Comment (toggle007) [reply test/webview/SampleView.test.ts]: Fixed

File: src/webview/SampleView.ts [hunk 5]
```diff
@@ -135,14 +136,41 @@ export class SampleView extends View {
...
+                const clonedRepo: RepositoryRegistryEntry = repositoryRegistry.get(options.repositoryName);
+                const repoPath: string = clonedRepo.path;
+                const readMeOnJSON: string = options.sample.readme;
+                const splitPath: string[] = readMeOnJSON.split(options.repositoryName);
+
+                const branchedPath: string = splitPath[1];
+                const endOfBranchName: number = branchedPath.indexOf('/', 1);
+                const slicedPath: string = branchedPath.slice(endOfBranchName + 1, branchedPath.length);
+
+                const readMe: string = await fs.readFileSync(path.join(repoPath, slicedPath), 'utf8');
```
Comment (cazfletch) [inline src/webview/SampleView.ts]: this should be `fs.readFile(.....` you don't need to use the synchronous version and use await

File: test/webview/SampleView.test.ts [hunk 6]
```diff
@@ -273,6 +292,35 @@ describe('SampleView', () => {
...
+
+            const error: Error = new Error('error happened');
+            mySandBox.stub(Axios, 'get').rejects(error);
+            mySandBox.stub(ejs, 'renderFile').callThrough();
+
+            const readFileSyncStub: sinon.SinonStub = mySandBox.stub(fs, 'readFileSync').resolves('README read here');
+
+            await RepositoryRegistry.instance().clear();
+
+            const repoEntry: RepositoryRegistryEntry = new RepositoryRegistryEntry();
```
Comment (cazfletch) [inline test/webview/SampleView.test.ts]: i thought you were going to put this in the beforeEach

Comment (toggle007) [reply test/webview/SampleView.test.ts]: I was planning to, but its only used twice - is it still worth doing it?

File: src/webview/SampleView.ts (hunk 5 above)
Comment (toggle007) [reply src/webview/SampleView.ts]: Will change that. Thanks

Comment (toggle007) [reply src/webview/SampleView.ts]: Fixed

File: test/webview/SampleView.test.ts (hunk 6 above)
Comment (cazfletch) [reply test/webview/SampleView.test.ts]: yes to get rid of the `repositoryGetStub` thats used later

Comment (toggle007) [reply test/webview/SampleView.test.ts]: Fixed - Removed the repositoryRegistryGetStub

**Pull request merged**